### Сортировка списков

1) Выучить наизусть сортировку пузырьком и быструю сортировку.


# 1.1. Сортировка пузырьком.(Bubble Sort)

In [5]:
def bubble_sort(arr) -> list:
    n = len(arr)
    for i in range(n-1):
        swapped = False 
        for j in range(n-1-i):
            if arr[j] > arr[j+1]:
                arr[j],arr[j+1] = arr[j+1],arr[j] 
                swapped = True
                
        if not swapped:
            break # Выход из цикла, если не было обменов
            
    return arr

bubble_sort([8,5,1,4,7])
                         

[1, 4, 5, 7, 8]

# 1.2. Быстрая сортировка.(Quick Sort)

In [6]:
import random 

def quick_sort(arr) -> list:
    if len(arr) <= 1:
        return arr
    else:
        pivot = random.choice(arr)
        left = [i for i in arr if i < pivot]
        centre = [i for i in arr if i == pivot]
        right = [i for i in arr if i > pivot]
    return quick_sort(left) + centre + quick_sort(right)

quick_sort([8,5,1,4,7])

[1, 4, 5, 7, 8]

2) Реализовать дополнительно 2 любые сортировки кроме вышеуказанных. Написать тесты на эти функции.


# 2.1. Сортировка вставками (Insertion Sort)

In [7]:
def insertion_sort(arr) -> list:
    n = len(arr)
    if len(arr) <= 1:
        return arr
    else:
        for i in range(1,n): #проходит до конца со второго(1) элемента массива
            for j in range(i, 0, -1): #j стартует от элемнета i и в обратном направлении пробегает массив
                if arr[j] < arr[j-1]:
                     arr[j], arr[j-1] = arr[j-1], arr[j]
                else:
                    break
    return arr

insertion_sort([8,5,1,4,7])

[1, 4, 5, 7, 8]

# 2.2. Сортировка выбором (Selection Sort)

In [8]:
def selection_sort(arr) -> list:
    n = len(arr)
    for i in range(n-1):
        min_idx = i # предполагаем, что текущий индекс содержит минимальный элемент
        for j in range(i+1,n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        if min_idx != i:
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr 
            
selection_sort([8,5,1,4,7])               

[1, 4, 5, 7, 8]

3) Создать список из 10000 случайных чисел с плавающей запятой.
4) Замерить и вывести время выполнения, пузырьком, быстрой сортировки, ваших 2 сортировок и встроенной сортировки python.


In [12]:
import random
import time

random_numbers = [random.uniform(-10000, 80000) for _ in range(10000)]


def measure_sort_time(sort_func, arr):
    'функция для измерения времени сортировки'
    start_time = time.perf_counter()
    sort_func(arr.copy())
    end_time = time.perf_counter()
    return end_time - start_time

bubble_sort_time = measure_sort_time(bubble_sort, random_numbers)
print(f'Время выполнения пузырьковой сортировки: {bubble_sort_time}')

quick_sort_time = measure_sort_time(quick_sort, random_numbers)
print(f'Время выполнения быстрой сортировки: {quick_sort_time}')

insertion_sort_time = measure_sort_time(insertion_sort, random_numbers)
print(f'Время выполнения сортировки вставками: {insertion_sort_time}')

selection_sort_time = measure_sort_time(selection_sort, random_numbers)
print(f'Время выполнения сортировки выбором: {selection_sort_time}')


Время выполнения пузырьковой сортировки: 6.123228222131729
Время выполнения быстрой сортировки: 0.023336709942668676
Время выполнения сортировки вставками: 3.858578613959253
Время выполнения сортировки выбором: 2.537172146141529


5) Создать список из 20000 случайных чисел с плавающей запятой. Сравните как изменилось время выполнения. 
(если на вашем компьютере это выполняется слишком долго больше минуты или не хватает памяти возьмите меньшее число элементов, например 2000 / 4000)


In [13]:
import random

random_numbers_2 = [random.uniform(-10000, 80000) for _ in range(20000)]

bubble_sort_time = measure_sort_time(bubble_sort, random_numbers_2)
print(f'Время выполнения пузырьковой сортировки: {bubble_sort_time}')

quick_sort_time = measure_sort_time(quick_sort, random_numbers_2)
print(f'Время выполнения быстрой сортировки: {quick_sort_time}')

insertion_sort_time = measure_sort_time(insertion_sort, random_numbers_2)
print(f'Время выполнения сортировки вставками: {insertion_sort_time}')

selection_sort_time = measure_sort_time(selection_sort, random_numbers_2)
print(f'Время выполнения сортировки выбором: {selection_sort_time}')


Время выполнения пузырьковой сортировки: 31.07508710795082
Время выполнения быстрой сортировки: 0.05862255999818444
Время выполнения сортировки вставками: 18.79863842506893
Время выполнения сортировки выбором: 11.787069213110954


6) Сделайте выводы по результатам, они достаточно интересные. Если будут вопросы спросите преподавателя.

При увеличении нагрузки в 2 раза время обработок сортировок изменилось следующим образом:
пузырьковая - время обработки увеличилось в 5 раз 
быстрая - время обработки увеличилось в 2.5 раза  
вставками - время обработки увеличилось в 5 раз 
выбором - время обработки увеличилось в 4.5 раза 

Медленее всего растет время затрат для быстрой сортировки, а значит она лучшая для сортировки больших массивов.
самая неэффективная для обработки больших объемов пузырьковая, так как время затрат выше, чем у всех остальных сортировок.